rclpy/src/rclpy_common/include/rclpy_common/handle.h

```diff
@@ -0,0 +1,116 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLPY_COMMON__HANDLE_H_
+#define RCLPY_COMMON__HANDLE_H_
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#include <Python.h>
+#include <stddef.h>
+
+#include "rcutils/types/rcutils_ret.h"
+
+/// Wrapper that manages the lifetime of an object,
+/// allowing to establish dependencies between them.
+typedef struct rclpy_handle_t rclpy_handle_t;
+
+/// Signature of rclpy_handle_t destructors.
+typedef void (* rclpy_handle_destructor_t)(void *);
+
+/// Create a PyCapsule wrapping a rclpy_handle_t object.
+/**
+ * Handles are used to manage lifetime of rcl objects.
+ * The main reason to use them, is that destruction order can't be controlled in Python.
+ * From PEP 442:
+ * > Cyclic isolate (CI)
+ * > A standalone subgraph of objects in which no object is referenced from the outside,
+ * > containing one or several reference cycles, and whose objects are still in a usable,
+ * > non-broken state: they can access each other from their respective finalizers.
+ * >
+ * > For CI objects, the order in which finalizers are called (step 2 above) is undefined.
+ *
+ * Handles provide a basic way of establishing dependencies and keeping the reference count,
+ * with a mechanism completly decoupled of Python reference counting.
```

Comment:
@ivanpauno nit: 
```suggestion
 * separately from Python's, ensuring proper destruction order. No protection against reference cycles is provided.
```

Reply:
Done in https://github.com/ros2/rclpy/pull/497/commits/269486e2444792a90bd0dc5c6e7a57c863272f6f.